flink-table/flink-table-planner-blink/src/main/java/org/apache/flink/table/planner/plan/nodes/exec/batch/BatchExecMultipleInput.java

```diff
@@ -0,0 +1,119 @@
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.flink.table.planner.plan.nodes.exec.batch;
+
+import org.apache.flink.api.dag.Transformation;
+import org.apache.flink.streaming.api.operators.ChainingStrategy;
+import org.apache.flink.streaming.api.transformations.MultipleInputTransformation;
+import org.apache.flink.table.data.RowData;
+import org.apache.flink.table.planner.delegation.BatchPlanner;
+import org.apache.flink.table.planner.plan.nodes.exec.ExecEdge;
+import org.apache.flink.table.planner.plan.nodes.exec.ExecNode;
+import org.apache.flink.table.planner.plan.nodes.exec.utils.ExecNodeUtil;
+import org.apache.flink.table.runtime.operators.multipleinput.BatchMultipleInputStreamOperatorFactory;
+import org.apache.flink.table.runtime.operators.multipleinput.TableOperatorWrapperGenerator;
+import org.apache.flink.table.runtime.operators.multipleinput.input.InputSpec;
+import org.apache.flink.table.runtime.typeutils.InternalTypeInfo;
+
+import org.apache.commons.lang3.tuple.Pair;
+
+import java.util.ArrayList;
+import java.util.List;
+import java.util.stream.Collectors;
+
+/**
+ * Batch exec node for multiple input which contains a sub-graph of {@link ExecNode}s.
+ * The root node of the sub-graph is {@link #outputNode}, and the leaf nodes of the sub-graph are
+ * the output nodes of the {@link #getInputNodes()}.
+ *
+ * <p>The following example shows a graph of {@code ExecNode}s with multiple input node:
+ * <pre>{@code
+ *          Sink
+ *           |
+ * +---------+--------+
+ * |         |        |
+ * |       Join       |
+ * |     /     \      | BatchExecMultipleInput
+ * |   Agg1    Agg2   |
+ * |    |       |     |
+ * +----+-------+-----+
+ *      |       |
+ * Exchange1 Exchange2
+ *      |       |
+ *    Scan1   Scan2
+ * }</pre>
+ *
+ * <p>The multiple input node contains three nodes: `Join`, `Agg1` and `Agg2`.
+ * `Join` is the root node ({@link #outputNode}) of the sub-graph,
+ * `Agg1` and `Agg2` are the leaf nodes of the sub-graph,
+ * `Exchange1` and `Exchange2` are the input nodes.
+ */
+public class BatchExecMultipleInput extends BatchExecNode<RowData> {
+
+	private final ExecNode<?> outputNode;
+
+	public BatchExecMultipleInput(
+			List<ExecNode<?>> inputNodes,
+			List<ExecEdge> inputEdges,
+			ExecNode<?> outputNode,
```

Comment:
it is a bit of confusing to sue outputNode here, the outputNode means the tail node of the sub graph instead of the down-stream node of BatchExecMultipleInput. May be keep it as tailNode would be better.

Reply:
In operator chaining, `tailNode` is appropriate, because the operators are list. While in multiple, the operators are graph (more strictly, it is a tree). so what about` rootNode` ?

Reply:
Nope, in multiple input operators the sub-operators can form a DAG, not necessarily a tree.

Reply:
yes, multiple input could support DAG, but currently it only support tree.